torch/distributed/_cute_layout.py

```diff
@@ -0,0 +1,318 @@
+                assert curr_size % sub_stride == 0 or sub_stride % curr_size == 0, (
+                    "Layouts do not meet stride divisibility condition"
+                )
+                new_size = min(max(1, curr_size // sub_stride), sub_size)
+                if new_size != 1:
+                    sub_res_sizes_and_strides.append(
+                        (new_size, sub_stride * curr_stride)
+                    )
+                assert sub_size % new_size == 0, (
+                    "Layouts do not meet shape divisibility condition"
+                )
+                sub_size = sub_size // new_size
+                sub_stride = self.ceil_div(sub_stride, curr_size)
+
+            # When self is integral and has single-size sublayout, aka, self = a:b, layout = s:d,
+            # the result is rather trivial: self o layout = a:b o s:d = s:(b*d).
+            # For example, if self = (6:2), layout = (3:2), the result is (3:(2*2)) = (3:4).
+            if sub_size != 1 or len(sub_res_sizes_and_strides) == 0:
+                sub_res_sizes_and_strides.append(
+                    (sub_size, sub_stride * self.strides[-1])
+                )
+
+            sub_res = _Layout(tuple(sub_res_sizes_and_strides))
+            res.append(sub_res)
+        return res
+
+    def complement(self, world_size: int) -> "_Layout":
+        """
+        Compute the "complement layout" relative to a given world_size.
+        A complement layout fills in the "missing" factor so that: self ⊗ complement(self, world_size)
+        has a total cosize (product of sizes) = world_size.
+
+        Example:
+          self = (4:1)   # size=4, stride=1
+          world_size = 8
+          Then:
+            cosize(self) = 4
+            Needed factor = 8 / 4 = 2
+            complement(self, 8) = (2:1)
+
+          Together they form:
+            (4:1) ⊗ (2:1) = (4,2):(2,1)
+          which has cosize = 4 * 2 = 8, as required.
+
+        In distributed terms, complement() is often used to derive the "other"
+        rank grouping when splitting processes into 2D meshes.
+        """
+        res_sizes_and_strides = []
+        current_idx = 1
+        for size, stride in sorted(self.sizes_and_strides, key=lambda x: x[1]):
+            if stride == 0 or size == 1:
+                continue
+            assert current_idx <= size * stride, (
+                f"current_idx {current_idx} larger than numel so far {size * stride}."
+            )
+
+            res_sizes_and_strides.append((stride // current_idx, current_idx))
+            current_idx = size * stride
+
+        res_sizes_and_strides.append(
+            (self.ceil_div(world_size, current_idx), current_idx)
+        )
+        # This is different from original pycute implementation, because in pytorch we usually
+        # have the right-most dimension as the innermost dimension (smallest stride).
+        res_sizes_and_strides.reverse()
```

Comment:
This is very, very important!!!!

Reply:
I don't think it's really accurate to call this CuTe layout. Maybe CuTe-inspired layout. But linearization being lexicographic rather than co-lexicographic by default is a really big difference and means most of the CuTe documentation cannot be used as is, because it's going the wrong way.

Reply:
ok maybe I can call it "_mesh_layout" instead?

Reply:
@tianyu-l I am not sure I am following here, we didn't make this assumption. So a layout can contain more than one size and stride, we never make assumption or say that "it seems to suggest it's enough for DeviceMesh to use depth==1 CuTe".